### PR TITLE
Relaxed CodeLocationsTest and co.

### PR DESCRIPTION
**Pull request by JaroslavTulach**
### Pull Request Description

Another set of fixes extracted from #3611. `CodeLocationsTest` test has been made more flexible to allow _"off by one"_ differences in the offset locations between the old and new parser.

### Checklist

Please include the following checklist in your PR:

- [x] All code conforms to the [Scala](https://github.com/enso-org/enso/blob/develop/docs/style-guide/scala.md), [Java](https://github.com/enso-org/enso/blob/develop/docs/style-guide/java.md), style guides.
- All code has been tested:
  - [x] Unit tests have been written where possible.

### CODE REVIEW

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 1]
```diff
@@ -162,11 +159,14 @@ IR.Module translateModule(Tree module) {
         var args = translateArgumentsDefinition(fn.getArgs());
         var body = translateExpression(fn.getBody());
 
+        if (body == null) {
+            throw new NullPointerException();
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:163]: Maybe throw a more informative exception? This seems like an internal error, no?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 2]
```diff
@@ -200,15 +200,21 @@ methodRef, args, def, getIdentifiedLocation(inputAst), meta(), diag()
       }
       case Tree.Assignment a -> {
         var reference = translateMethodReference(a.getPattern(), false);
+        var body = translateExpression(a.getExpr());
+        if (body == null) {
+            throw new NullPointerException();
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:205]: same comment as in the previous case

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 3]
```diff
@@ -528,13 +565,24 @@ yield switch (identifier) {
...
+              default -> getIdentifiedLocation(tree);
+            }, body.location());
+            yield new IR$Function$Lambda(args, body, at, true, meta(), diag());
           }
           default -> {
             var lhs = unnamedCallArgument(app.getLhs());
             var rhs = unnamedCallArgument(app.getRhs());
+            if ("@".equals(op.codeRepr()) && lhs.value() instanceof IR$Application$Prefix fn) {
+                final Option<IdentifiedLocation> where = getIdentifiedLocation(op);
+                var err = new IR$Error$Syntax(where.get(), IR$Error$Syntax$UnrecognizedToken$.MODULE$, meta(), diag());
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:579]: do we check/ensure somewhere that `where` is non-empty?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 4]
```diff
@@ -610,7 +659,13 @@ yield switch (identifier) {
           expressions.remove(expressions.size()-1);
         }
         var list = CollectionConverters.asScala(expressions.iterator()).toList();
-        yield new IR$Expression$Block(list, last, getIdentifiedLocation(body), false, meta(), diag());
+        var loc = getIdentifiedLocation(body, 0, 1);
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: can you add some comments on those magic constants?

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java [hunk 5]
```diff
@@ -651,7 +709,13 @@ yield switch (fn) {
         }
       }
       case Tree.TypeAnnotated anno -> translateTypeAnnotated(anno);
-      case Tree.Group group -> translateExpression(group.getBody(), false);
+      case Tree.Group group -> {
+          yield switch (translateExpression(group.getBody(), false)) {
+              case null -> new IR$Error$Syntax(getIdentifiedLocation(group).get(), IR$Error$Syntax$EmptyParentheses$.MODULE$, meta(), diag());
+              case IR$Application$Prefix pref -> pref.setLocation(getIdentifiedLocation(group, 1, -1));
```
Comment (hubertp) [inline engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: can you add a comment on the magic `1, -1` constants?

**Review by hubertp — COMMENTED** [review]
I think it would be useful to have comments in various places where locations are manipulated.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 1 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:163]: The information is the stack trace!

Previously the execution continued and later, when certain compiler phase tried to traverse the `IR`, another `NullPointerException` was thrown. That one was however non-informative. Throwing the exception early on helps to identify the source of the problem.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 2 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:205]: As above. The stacktrace is the information.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 3 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java:579]: `getIdentifiedLocation` always returns non-empty option for non-null `op`. The code `op.codeRepr()` on previous line ensures `op != null`.

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 5 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: OK. Done in 081c79700

File: engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java (hunk 4 above)
Comment (JaroslavTulach) [reply engine/runtime/src/main/java/org/enso/compiler/TreeToIr.java]: OK, done in [ea1a895](https://github.com/enso-org/enso/pull/3849/commits/ea1a895e583bf706a58e1bce39fd31d2767a8e99)

**Comment by JaroslavTulach** [comment]
> I think it would be useful to have comments in various places where locations are manipulated.

OK, I've just renamed the local variables to describe their purpose. Btw. the cases are covered by associated unit tests.

**Review by hubertp — APPROVED** [review]

**Pull request merged**